src/storage/src/source/kafka.rs

```diff
@@ -303,27 +330,32 @@ impl SourceReader for KafkaSourceReader {
     ///
     /// If a message has an offset that is smaller than the next expected offset for this consumer
     /// (and this partition) we skip this message, and seek to the appropriate offset
-    fn get_next_message(&mut self) -> NextMessage<Self::Key, Self::Value, Self::Diff> {
+    fn get_next_message(&mut self) -> NextMessage<Self::Key, Self::Value, Self::Time, Self::Diff> {
         let partition_info = self.partition_info.lock().unwrap().take();
         if let Some(partitions) = partition_info {
-            // NOTE: We're somewhat inefficient with Vec allocations and the
-            // like. Shouldn't be a problem though, because we rarely hear about
-            // new partitions.
-            let mut unconsumed_partitions = Vec::new();
+            let mut max_pid = None;
             for pid in partitions {
-                let pid = PartitionId::Kafka(pid);
-                if crate::source::responsible_for(&self.id, self.worker_id, self.worker_count, &pid)
-                {
+                max_pid = std::cmp::max(max_pid, Some(pid));
+                let is_responsible = crate::source::responsible_for(
+                    &self.id,
+                    self.worker_id,
+                    self.worker_count,
+                    pid,
+                );
+                if is_responsible {
                     self.ensure_partition(pid);
-                } else {
-                    unconsumed_partitions.push(pid);
+                    let part_min_ts = Partitioned::with_partition(pid, MzOffset::from(0));
+                    self.partition_data_capabilities
+                        .entry(pid)
+                        .or_insert_with(|| self.data_capability.delayed(&part_min_ts));
+                    self.partition_upper_capabilities
+                        .entry(pid)
+                        .or_insert_with(|| self.upper_capability.delayed(&part_min_ts));
                 }
             }
-            if !unconsumed_partitions.is_empty() {
-                return NextMessage::Ready(SourceMessageType::DropPartitionCapabilities(
-                    unconsumed_partitions,
-                ));
-            }
+            let future_ts = Partitioned::with_range(max_pid, None, MzOffset::from(0));
+            self.data_capability.downgrade(&future_ts);
+            self.upper_capability.downgrade(&future_ts);
```

Comment:
I re-thought about this, and realized I was thinking about it wrong (actually I was wrong about how timely worked, and I forgot that partitions cannot be deleted). Usually this means something is subtle enough that is warrants a comment:

Perhaps, at the struct definitions for these fields, we say something like (please carefully fact check me):

```
/// The `data_capability` and `upper_capability` are directly derived from the initialization capabilities communicated to us in `into_reader`, by the source reader pipeline.
/// We maintain capabilities at `MzOffset(0)` for the range of all partitions greater than the max partition managed by the current worker. Note that because kafka partitions can _never_ be delete, this is sufficient. We manage and maintain this in `into_reader` during initialization, and when the partition set managed by the worker changes.
/// We use these capabilities to mint new-per-partition capabilities that are downgraded as we produce messages for that partition.
/// Note that this means that _always_ maintain a capability at `MzOffset(0)` for some `Interval::Range` timestamp. While this may _seem_ like it prevents the ability of the `source_reader_operator` to progress its frontier, but, critically, timely internally maintains a `MutableAntichain` that is updated as we downgrade the derived, per-partition capabilities and emit messages. This Antichain will contain a `Point` `Partitioned` for each partition, as well as a `(Range(Top, Bottom), 0)` that covers partitions for which there are no workers.
/// Interestingly, and perhaps _fundamentally_, this `MutableAntichain` has a functionally equivalent _dual_ which is maintained by us at the read end of a `Capture` in the `reclock_operator` in the `source_reader_operator` module!
```

that last sentence maybe should go somewhere in `source_reader_operator.rs` tho...

Also, this brings up an interesting point: what will the antichain we perist in the reclock shard look like? I dont think its going to be nice and neat `Point`'s with `Range`'s in between...its gonna be one big range with a bunch of points scattered about, right?